### ISSUE TITLE
panic: Daemon panics if started with no internet connection on Windows

### ISSUE THREAD

**Issue by mateon1** [ipfs/go-ipfs#3226]
#### Version information:
go-ipfs version: 0.4.3-rc2-
Repo version: 4
System version: amd64/windows
Golang version: go1.6.2

#### Type:
bug

#### Priority:
P2

#### Description:
If the daemon is started while no internet connection is available (e.g. wifi disconnected), the daemon panics while trying to 'work around NAT', this panic is fully reproducible on my machine.

```
10:36:11.723  INFO   cmd/ipfs: IPFS_PATH D:\ipfs main.go:297
10:36:11.723 DEBUG   cmd/ipfs: Command cannot run on daemon. Checking if daemon is locked main.go:425
10:36:11.723 DEBUG       lock: Checking lock lock.go:32
10:36:11.724 DEBUG       lock: No one has a lock lock.go:61
10:36:11.724 DEBUG   cmd/ipfs: calling pre-command hooks... main.go:291
10:36:11.724 DEBUG   cmd/ipfs: executing command locally main.go:338
Initializing daemon...
10:36:11.835 DEBUG       core: Config.Addresses.Swarm:[/ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] core.go:596
10:36:11.835 DEBUG       core: Config.Addresses.Swarm:[/ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] (filtered) core.go:598
10:36:11.835 DEBUG       conn: Conn Listener on /ip4/0.0.0.0/tcp/4001 listen.go:222
10:36:11.835 DEBUG     swarm2: Swarm Listeners at [/ip4/0.0.0.0/tcp/4001] swarm_listen.go:102
10:36:11.835 DEBUG       conn: Conn Listener on /ip6/::/tcp/4001 listen.go:222
10:36:11.835 DEBUG     swarm2: Swarm Listeners at [/ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] swarm_listen.go:102
10:36:11.840 DEBUG github.com: InterfaceAddresses: from manet:[/ip6/fe80::cc6:9801:91ad:3546 /ip4/169.254.53.70 /ip6/fe80::a5a1:ceb2:6176:89f2 /ip4/169.254.137.242 /ip6/fe80::6c16:a991:9cdb:e45 /ip4/169.254.14.69 /ip6/fe80::24d4:a611:e00a:4fe3 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1 /ip6/fe80::ffff:ffff:fffe] addr.go:204
10:36:11.840 DEBUG github.com: InterfaceAddresses: usable:[/ip4/169.254.53.70 /ip4/169.254.137.242 /ip4/169.254.14.69 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1] addr.go:216
10:36:11.840 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.53.70/tcp/4001 [/ip4/169.254.53.70/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.137.242/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.14.69/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.79.227/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/127.0.0.1/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001 /ip4/127.0.0.1/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: adding resolved addr:/ip6/::/tcp/4001 /ip6/::1/tcp/4001 [/ip6/::1/tcp/4001] addr.go:147
10:36:11.840 DEBUG github.com: ResolveUnspecifiedAddresses:[/ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] [/ip4/169.254.53.70 /ip4/169.254.137.242 /ip4/169.254.14.69 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1] [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001 /ip4/127.0.0.1/tcp/4001 /ip6/::1/tcp/4001] addr.go:192
10:36:11.840  INFO       core: Swarm listening at: [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001 /ip4/127.0.0.1/tcp/4001 /ip6/::1/tcp/4001] core.go:613
10:36:11.842 DEBUG github.com: InterfaceAddresses: from manet:[/ip6/fe80::cc6:9801:91ad:3546 /ip4/169.254.53.70 /ip6/fe80::a5a1:ceb2:6176:89f2 /ip4/169.254.137.242 /ip6/fe80::6c16:a991:9cdb:e45 /ip4/169.254.14.69 /ip6/fe80::24d4:a611:e00a:4fe3 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1 /ip6/fe80::ffff:ffff:fffe] addr.go:204
10:36:11.842 DEBUG github.com: InterfaceAddresses: usable:[/ip4/169.254.53.70 /ip4/169.254.137.242 /ip4/169.254.14.69 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1] addr.go:216
10:36:11.842 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.53.70/tcp/4001 [/ip4/169.254.53.70/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.137.242/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.14.69/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/169.254.79.227/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: adding resolved addr:/ip4/0.0.0.0/tcp/4001 /ip4/127.0.0.1/tcp/4001 [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001 /ip4/127.0.0.1/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: adding resolved addr:/ip6/::/tcp/4001 /ip6/::1/tcp/4001 [/ip6/::1/tcp/4001] addr.go:147
10:36:11.842 DEBUG github.com: ResolveUnspecifiedAddresses:[/ip4/0.0.0.0/tcp/4001 /ip6/::/tcp/4001] [/ip4/169.254.53.70 /ip4/169.254.137.242 /ip4/169.254.14.69 /ip4/169.254.79.227 /ip6/::1 /ip4/127.0.0.1] [/ip4/169.254.53.70/tcp/4001 /ip4/169.254.137.242/tcp/4001 /ip4/169.254.14.69/tcp/4001 /ip4/169.254.79.227/tcp/4001 /ip4/127.0.0.1/tcp/4001 /ip6/::1/tcp/4001] addr.go:192
10:36:11.844 DEBUG       core: <peer.ID T4Lp3H> bootstrapping to 4 nodes: [{<peer.ID SoLPpp> [/ip4/104.236.179.241/tcp/4001]} {<peer.ID aCpDMG> [/ip4/104.131.131.82/tcp/4001]} {<peer.ID SoLueR> [/ip4/162.243.248.213/tcp/4001]} {<peer.ID SoLV4B> [/ip4/104.236.76.40/tcp/4001]}] bootstrap.go:149
10:36:11.846 DEBUG       core: <peer.ID T4Lp3H> bootstrapping to <peer.ID SoLV4B> bootstrap.go:174
10:36:11.846 DEBUG       core: <peer.ID T4Lp3H> bootstrapping to <peer.ID aCpDMG> bootstrap.go:174
10:36:11.846 DEBUG       core: <peer.ID T4Lp3H> bootstrapping to <peer.ID SoLueR> bootstrap.go:174
10:36:11.846 DEBUG github.com: host %!s(func() peer.ID=0x7d0470) dialing <peer.ID aCpDMG> basic_host.go:201
10:36:11.846 DEBUG github.com: host %!s(func() peer.ID=0x7d0470) dialing <peer.ID SoLueR> basic_host.go:201
10:36:11.846 DEBUG     swarm2: [<peer.ID T4Lp3H>] network dialing peer [<peer.ID aCpDMG>] swarm_net.go:36
10:36:11.846 DEBUG     swarm2: [<peer.ID T4Lp3H>] network dialing peer [<peer.ID SoLueR>] swarm_net.go:36
10:36:11.846 DEBUG github.com: host %!s(func() peer.ID=0x7d0470) dialing <peer.ID SoLV4B> basic_host.go:201
10:36:11.847 DEBUG     swarm2: [<peer.ID T4Lp3H>] network dialing peer [<peer.ID SoLV4B>] swarm_net.go:36
10:36:11.846 DEBUG       core: <peer.ID T4Lp3H> bootstrapping to <peer.ID SoLPpp> bootstrap.go:174
10:36:11.849 DEBUG github.com: host %!s(func() peer.ID=0x7d0470) dialing <peer.ID SoLPpp> basic_host.go:201
panic: runtime error: index out of range

goroutine 113 [running]:
panic(0xe36ca0, 0xc082008020)
        /home/whyrusleeping/go/src/runtime/panic.go:481 +0x3f4
gx/ipfs/Qma5KQppqMBmpdNeVUDuNtEkec1BRNWcS3dodUdiP9267q/gateway.DiscoverGateway(0x0, 0x0, 0x0, 0x0, 0x0)
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/Qma5KQppqMBmpdNeVUDuNtEkec1BRNWcS3dodUdiP9267q/gateway/gateway_windows.go:37 +0x4e6
gx/ipfs/QmNLvkCDV6ZjUJsEwGNporYBuZdhWT6q7TBVYQwwRv12HT/go-nat.discoverNATPMP(0xc083c8c180)
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmNLvkCDV6ZjUJsEwGNporYBuZdhWT6q7TBVYQwwRv12HT/go-nat/natpmp.go:18 +0x46
gx/ipfs/QmNLvkCDV6ZjUJsEwGNporYBuZdhWT6q7TBVYQwwRv12HT/go-nat.DiscoverGateway(0x0, 0x0, 0x0, 0x0)
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmNLvkCDV6ZjUJsEwGNporYBuZdhWT6q7TBVYQwwRv12HT/go-nat/nat.go:44 +0x86
gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/nat.DiscoverNAT(0x8)
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/nat/nat.go:37 +0x2b
gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/host/basic.(*natManager).discoverNAT.func1.1(0xc083c8c0c0, 0xc082026060)
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/host/basic/natmgr.go:81 +0x46
created by gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/host/basic.(*natManager).discoverNAT.func1
        /builds/distributions/dists/go-ipfs/gopath/src/gx/ipfs/QmVCe3SNMjkcPgnpFhZs719dheq6xE7gJwjzV7aWcUM4Ms/go-libp2p/p2p/host/basic/natmgr.go:82 +0x96
```

`ipfs daemon --offline` launches fine.

**Comment by Kubuxu** [comment]
Can you try the 0.4.4-dev from the master? We have updated the go-nat which might have fixed the issue.

If you can build it, try it. If you can't I can build it for you.

**Comment by mateon1** [comment]
@Kubuxu Okay, I am currently installing Go 1.7.1 on my laptop. Will report soon

**Comment by mateon1** [comment]
This is fixed in `0.4.4-dev-73cd8b3`.
Sorry for the delay, but I had a really hard time dealing with gx repo corruption, due to my school network having an impenetrable NAT, which caused ipfs downloads to fail.
Both of these seem like bugs, but I'm too tired at the moment to reproduce them.
